Feature/export page cleanup 2

AdityaKhatri opened the pull request:
depends on https://github.com/the-deep/server/pull/1104
## Changes

* Add auto-save functionality to filters
* Add clone-able export

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] codegen errors
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [x] translations

samshara committed:
Save project sources filters in server

samshara committed:
Fetch sources filters and transform to form acceptable value

samshara committed:
Reset active page when sources filter change

samshara committed:
Add reapply previous export filters to create new export

samshara committed:
Fix type issues and improve styling in export history page

samshara committed:
Add Apply previous filters button

AdityaKhatri committed:
Save sources filter after lead/entry data is fetched

tnagorra commented on app/views/Project/Tagging/Export/ExportHistory/TableActions/index.tsx:
```diff
@@ -51,6 +52,44 @@ function TableActions(props: Props) {
+                    title="Create new export with these filters"
+                    route={routes.exportCreate}
+                    attrs={{
+                        projectId: data.project as string | undefined,
```
Any reason we need to cast this type?

tnagorra commented on app/views/Project/Tagging/Export/ExportHistory/TableActions/index.tsx:
```diff
@@ -51,6 +52,44 @@ function TableActions(props: Props) {
+                    title="Create new assessment export with these filters"
+                    route={routes.assessmentExportCreate}
+                    attrs={{
+                        projectId: data.project as string | undefined,
```
Any reason we need to cast this type?

tnagorra commented on app/views/Project/Tagging/Export/ExportHistory/TableActions/styles.css:
```diff
@@ -2,6 +2,19 @@
+            display: flex;
+            justify-content: center;
+            padding: 0;
+            width: 1.5em;
+            height: 1.5em;
+            font-size: 1.2em;
```
Why are we directly using `em` instead of design tokens?

samshara replied:
@AdityaKhatri

tnagorra commented on app/views/Project/Tagging/Export/ExportHistory/index.tsx:
```diff
@@ -113,8 +110,6 @@ function ExportHistory(props: Props) {
     const [activePage, setActivePage] = useState(1);
     const [exportedAt, setExportedAt] = useState<DateRangeValue>();
     const [searchText, setSearchText] = useState<string>();
-    // FIXME: we can safely remove useDebouncedValue if we are using TextInput from deep-ui
-    const debouncedSearchText = useDebouncedValue(searchText, debounceTime);
```
Any reason we removed debouncing?

samshara replied:
The TextInput component supports debouncing

tnagorra commented on app/views/Project/Tagging/Export/NewAssessmentExport/index.tsx:
```diff
@@ -93,24 +82,32 @@ function NewAssessmentExport(props: Props) {
         setPristine,
     } = useFilterState();
 
+    useEffect(() => {
+        if (state?.filters) {
+            setSourcesFilterValue(transformRawFiltersToFormValues(state?.filters));
+        }
+    }, [setSourcesFilterValue, state?.filters]);
```
Does this render only on the first render?

samshara replied:
yes

tnagorra commented on app/views/Project/Tagging/Export/NewExport/index.tsx:
```diff
@@ -138,43 +128,75 @@ function NewExport(props: Props) {
         setPristine,
     } = useFilterState();
 
+    useEffect(() => {
+        if (state?.filters) {
+            setSourcesFilter(transformRawFiltersToFormValues(state?.filters));
+        }
+    }, [setSourcesFilter, state?.filters]);
```
Does this render only on the first render?

samshara replied:
yes

tnagorra commented on app/views/Project/Tagging/Export/NewExport/index.tsx:
```diff
@@ -324,45 +346,66 @@ function NewExport(props: Props) {
+                        ...v,
+                        selected: state?.extraOptions?.reportTextWidgetIds
+                            ? (state?.extraOptions
+                                ?.reportTextWidgetIds?.includes(v.id) ?? false)
```
```suggestion
                        selected: state?.extraOptions?.reportTextWidgetIds
                            ? (state.extraOptions.reportTextWidgetIds.includes(v.id) ?? false)
```

tnagorra commented on app/views/Project/Tagging/Export/NewExport/index.tsx:
```diff
@@ -324,45 +346,66 @@ function NewExport(props: Props) {
+                        selected: state?.extraOptions?.reportExportingWidgets
+                            ? (state?.extraOptions
+                                ?.reportExportingWidgets?.includes(v.id) ?? false)
+                            : true,
```
Similar change as the above comment.

tnagorra commented on app/views/Project/Tagging/Export/NewExport/index.tsx:
```diff
@@ -436,8 +479,8 @@ function NewExport(props: Props) {
     const isFilterEmpty = useMemo(() => (
-        doesObjectHaveNoData(sourcesFilter, ['', null])
-    ), [sourcesFilter]);
+        doesObjectHaveNoData(sourcesFilters, ['', null])
```
Is there any way we can have `null` here?

tnagorra commented on app/views/Project/Tagging/Export/utils.ts:
```diff
@@ -351,5 +352,37 @@ export const createWidgetIds = (widgets: TreeSelectableWidget[]) => (
+export const isSubSectorIncluded = (
+    reportStructure: ExportItem['extraOptions']['reportStructure'],
+) => (
+    reportStructure?.some((rootLevel) => (
+        rootLevel?.levels?.some((l) => (
+            l?.levels?.some((m) => (
+                m?.levels?.some((n) => (
+                    n?.levels?.some((o) => (
+                        isDefined(o?.id)
+                    ))
+                ))
+            ))
+        ))
```
Wut!

samshara replied:
this is one way to find if sub-sector is included

tnagorra commented on app/views/Project/Tagging/Export/utils.ts:
```diff
@@ -351,5 +352,37 @@ export const createWidgetIds = (widgets: TreeSelectableWidget[]) => (
+            const rowKeys = (
+                widget.properties as Matrix2dProperties
+            )?.rows.map((v) => v.key);
+            return reportStructure?.some((rootLevel) => (
+                rootLevel?.levels?.some((level) => rowKeys?.includes(level.id))
+            ));
```
We have a lot of nesting. I will need to get back to this later.

tnagorra commented on app/views/Project/Tagging/Sources/EntriesGrid/index.tsx:
```diff
@@ -254,6 +256,10 @@ function EntriesGrid(props: Props) {
     const [activePage, setActivePage] = useState(1);
     const [maxItemsPerPage, setMaxItemsPerPage] = useState(50);
 
+    useEffect(() => {
+        setActivePage(1);
+    }, [rawFilters]);
+
```
This is not how filters should be reset

samshara replied:
`rawFilters` are set in the parent component and each subsequent change in this must trigger a fetch while updating the `activePage` state.

tnagorra replied:
The problem here is that there will be two requests on the server

tnagorra commented on app/views/Project/Tagging/Sources/EntriesGrid/index.tsx:
```diff
@@ -293,6 +299,7 @@ function EntriesGrid(props: Props) {
                 if (!projectFromResponse) {
                     return;
                 }
+                onSourcesGetSuccess();
```
Let's move this at the end of the function

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/index.tsx:
```diff
@@ -127,9 +127,9 @@ function EntryFilter<K extends string>(props: Props<K>) {
-                        && doesObjectHaveNoData(value?.createdAtLte)
-                        && !allFiltersVisible
-                        && styles.hidden,
+                    && doesObjectHaveNoData(value?.createdAtLte)
+                    && !allFiltersVisible
+                    && styles.hidden,
```
IDK if this indentation is correct

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -140,19 +133,17 @@ function SourcesTable(props: Props) {
     const [selectedLeads, setSelectedLeads] = useState<Lead[]>([]);
     const [maxItemsPerPage, setMaxItemsPerPage] = useState(defaultMaxItemsPerPage);
 
+    useEffect(() => {
+        setActivePage(1);
+    }, [filters]);
```
This is not the correct way to clear active page.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -179,6 +170,9 @@ function SourcesTable(props: Props) {
         {
             skip: isNotDefined(variables),
             variables,
+            onCompleted: () => {
+                onSourcesGetSuccess();
+            },
```
We need to check if the reponse is okay

samshara replied:
this is a query

tnagorra replied:
Even if this is a query, there might be an error

tnagorra commented on app/views/Project/Tagging/Sources/index.tsx:
```diff
@@ -137,12 +211,41 @@ function Sources(props: Props) {
     }, [setError, validate, handleSubmit]);
 
+    const handleUsePreviousLeadFilter = useCallback(() => {
+        if (projectSavedLeadFilterData?.project) {
```
Let's early return instead of nesting in a `if` block

tnagorra commented on app/views/Project/Tagging/Sources/queries.ts:
```diff
@@ -0,0 +1,311 @@
+`;
+
+export const SOURCE_FILTER_FRAGMENT = gql`
+    fragment SourceFilterResponse on LeadsFilterDataType {
+        assignees
+        authorOrganizations
```
Please check if we have `id`

samshara replied:
no id is present

tnagorra commented on app/views/Project/Tagging/Sources/utils.tsx:
```diff
@@ -46,3 +53,48 @@ export function transformSourcesFilterToEntriesFilter(filters: FaramValues) {
+    return undefined;
+}
+
+function getDateString(dateTimeString: string | null | undefined) {
+    if (dateTimeString && Date.parse(dateTimeString)) {
+        return encodeDate(new Date(dateTimeString));
+    }
+    return dateTimeString;
+}
```
Why are we not using a utils from fujs?

samshara replied:
we need to check if `dateTimeString` is valid date string value.

tnagorra replied:
Why do we have such cases?

tnagorra commented on app/views/Project/Tagging/Sources/utils.tsx:
```diff
@@ -46,3 +53,48 @@ export function transformSourcesFilterToEntriesFilter(filters: FaramValues) {
+    orderingFilter: LeadOrderingEnum[] | null | undefined,
+): { name: string; direction: SortDirection } | undefined {
+    const [direction, ...orderingName] = orderingFilter?.[0] ? orderingFilter[0].split('_') : [];
+    if (direction) {
```
Let's early return

tnagorra commented on app/views/Project/Tagging/Sources/utils.tsx:
```diff
@@ -46,3 +53,48 @@ export function transformSourcesFilterToEntriesFilter(filters: FaramValues) {
+}
+
+export function transformRawFiltersToFormValues(filters: RawSourcesFilter) {
+    if (filters) {
```
Let's early return

samshara committed:
Fix pull request review comments on Export view

samshara committed:
Preserve report structure sort info when create new export with previous filters

samshara committed:
Add test cases for create export utility functions

tnagorra commented on app/views/Project/PillarAnalysis/SourcesFilter/EntryFilter/index.tsx:
```diff
@@ -118,9 +118,9 @@ function EntryFilter<K extends string>(props: Props<K>) {
-                        && doesObjectHaveNoData(value?.createdAtLte)
-                        && !allFiltersVisible
-                        && styles.hidden,
+                    && doesObjectHaveNoData(value?.createdAtLte)
+                    && !allFiltersVisible
+                    && styles.hidden,
```
Why was the whitespace changed?

tnagorra commented on app/views/Project/PillarAnalysis/SourcesFilter/types.ts:
```diff
@@ -21,6 +21,6 @@ export type SourcesFilterFields = PurgeNull<EnumFix<ProjectSourcesQueryVariables
 export type SourceFilterOptions = DeepReplace<
     SourceFilterOptionsQuery,
-    Omit<AnalysisFrameworkFilterTypeRaw, 'widgetTypeDisplay' | 'filterTypeDisplay'>,
+    Omit<AnalysisFrameworkFilterTypeRaw, 'widgetTypeDisplay' | 'filterTypeDisplay' | 'widgetKey'>,
```
Why omit `widgetKey`?

tnagorra committed:
Move clearing of entry filter from submission to filter change

- Fix filter for Select widget
- Filter out obsolete widgets and selected keys inside each widget
  filter
- Change logic of apply/clear visibility on lead filter
- Rename AppliedFilters to SourceAppliedFilters
- Break AppliedFilters into smaller components

tnagorra committed:
Flatten views heirarchy

- Move common fragments to gqlFragments file
- Move common components in components dir
- Add absolute imports for gqlFragments and redirects
- Remove unused typings
- Add madge config

AdityaKhatri commented on app/components/framework/CompactAttributeInput/OrganigramWidgetInput/utils.ts:
```diff
@@ -0,0 +1,20 @@
+import { OrganigramDatum } from '#types/newAnalyticalFramework';
+
+// FIXME: move this to utils
```
Remove this, as its already done. Or are you talking about another utils?

AdityaKhatri commented on app/components/leadFilters/SourcesAppliedFilters/EntryFilterOutput/index.tsx:
```diff
@@ -0,0 +1,152 @@
+                }
+                return (
+                    <FrameworkFilterOutput
+                        // FIXME: why are we capitalizing the framework filter title
```
We are doing this so has to make it consistent with rest of the filter labels and capitalize them all.

AdityaKhatri commented on app/components/leadFilters/SourcesFilter/EntryFilter/index.tsx:
```diff
@@ -195,26 +223,25 @@ function EntryFilter<K extends string>(props: Props<K>) {
+                        key={filter.id}
+                        variant="general"
+                        name={filterValue?.index}
+                        // FIXME: any reason to do this?
```
capitalizing everything

AdityaKhatri commented on app/components/leadFilters/SourcesFilter/types.ts:
```diff
@@ -0,0 +1,18 @@
+    | 'confidentiality'
+    | 'priorities'
+    | 'ordering'
+
```
remove empty line

AdityaKhatri commented on app/utils/common.ts:
```diff
@@ -133,7 +133,8 @@ export const enumKeySelector = <T>(d: EnumEntity<T>) => (
 
 export const enumLabelSelector = <T extends string | number>(d: EnumEntity<T>) => (
-    capitalize(d.description ?? `${d.name}`)
+    // FIXME: why are we doing this
```
For consistency in filter and filter options

AdityaKhatri reviewed (commented):
Some points

tnagorra committed:
Separate out components used in Export

AdityaKhatri committed:
Fix filter options in export page
